src/Surfnet/StepupGateway/GatewayBundle/Controller/GatewayController.php

```diff
@@ -191,14 +192,33 @@ public function sendAuthenticationCancelledByUserAction()
      */
     public function renderSamlResponse($view, SAMLResponse $response, $authenticationMode)
     {
+        $logger = $this->get('logger');
+        /** @var ResponseHelper $responseHelper */
+        $responseHelper = $this->get('second_factor_only.adfs.response_helper');
+
         $this->supportsAuthenticationMode($authenticationMode);
         $responseContext = $this->getResponseContext($authenticationMode);
 
-        return $this->render($view, [
-            'acu'        => $responseContext->getDestination(),
-            'response'   => $this->getResponseAsXML($response),
+        $parameters = [
+            'acu' => $responseContext->getDestination(),
+            'response' => $this->getResponseAsXML($response),
             'relayState' => $responseContext->getRelayState()
-        ]);
+        ];
+
+        // Test if we should add ADFS response parameters
+        $inResponseTo = $responseContext->getInResponseTo();
+        if ($responseHelper->isAdfsResponse($inResponseTo)) {
+            $adfsParameters = $responseHelper->retrieveAdfsParameters();
+            $logMessage = 'Responding with additional ADFS parameters, in response to request: "%s", with view: "%s"';
+            if ($response->isSuccess()) {
+                $logMessage = 'Responding with an AuthnFailed SamlResponse with ADFS parameters, in response to AR: "%s", with view: "%s"';
+            }
+            $logger->notice(sprintf($logMessage, $inResponseTo, $view));
+            $parameters['adfs'] = $adfsParameters;
+            $parameters['acu'] = $responseContext->getDestinationForAdfs();
+        }
+
+        return $this->render($view, $parameters);
```

Comment:
This might seem illogical. We are responding to an SFO SAML Response (Adfs authentications are intrinsically SFO auth) from the 'non-sfo' Gateway controller. 

That is because all AuthnFailed saml responses are redirected via that route. And that is why those endpoints now also support the ADFS response parameters.

Not my favourite, and feel free to discuss a possible better solution. For one, I reconsidered adding an Event Listener on the response sent event. We could add the additional ADFS parameters there for any ACS action that is performed. But that also raises the magic factor up a notch..

Reply:
Maybe consider a dedicated failed SAML response route to redirect to to handle failed authentications. Or is that not possible?
Not raising the level of magic however is certainly a good call!